### PR TITLE
Set GraphQlInputException when not formatted

### PR DESCRIPTION
**Pull request by sedonik**
### Description (*)

Added Magento\CatalogGraphQl\Model\Product\Option\DateTypePool class.
Added dataTypes array (Y-m-d, Y-m-d H:i:s, Y-m-d H:i:s) in di.xml.
Implemented DI for DateTypePool with the constructor.

### Fixed Issues

1. magento/graphql-ce#761: [Customizable Options] Call to a member function format() on boolean

### Manual testing scenarios (*)

1. Create Simple Product similar to :
SKU: simple-product-1
qty: 100500
Customizable options:
type: date

2. Perform query for Customer and use header
Header:

```
{
  "Authorization": "Bearer {{ CUSTOMER_TOKEN }}"
}
```

Query:

```
mutation {
  addSimpleProductsToCart(
    input: {
      cart_id: "{{ CART_ID }}"
      cart_items: {
        data: {
          quantity: 1
          sku: "simple-product-3"
        }
        customizable_options: [
          {
            id: 1
            value_string: "12-12-12"
          }
        ]
      }
    }
  ) {
    cart {
      items {
        quantity
        product {
          sku
        }
      }
    }
  }
}
```
3. Product was added to cart

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-5548](https://jira.corp.magento.com/browse/ENGCOM-5548) has been created to process this Pull Request
:eight_spoked_asterisk: @lenaorobei, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by lenaorobei** [comment]
@sedonik please notice static build failure. It says `/var/www/html/app/code/Magento/CatalogGraphQl/Model/Product/Option/DateType.php:18	The class Magento\CatalogGraphQl\Model\Product\Option\DateType uses sessions or cookies while not being a part of HTML Presentation layer` and it is caused by parent constructor. Since we cannot refactor this code (remove session from constructor) because of backward compatibility. In this case we need to add `@SuppressWarnings(PHPMD.CookieAndSessionMisuse)` to the `\Magento\CatalogGraphQl\Model\Product\Option\DateType` class like it is added to its parent `\Magento\Catalog\Model\Product\Option\Type\Date`.

**Comment by sedonik** [comment]
@lenaorobei  
Thank you for the notice.

I've added @SuppressWarnings(PHPMD.CookieAndSessionMisuse) to class Magento\CatalogGraphQl\Model\Product\Option\DateType

Pls, check.

File: app/code/Magento/CatalogGraphQl/Model/Product/Option/DateType.php [hunk 1]
```diff
@@ -10,9 +10,13 @@
 use Magento\Catalog\Model\Product\Option\Type\Date as ProductDateOptionType;
 use Magento\Framework\Exception\LocalizedException;
 use Magento\Framework\Stdlib\DateTime;
+use Magento\Framework\GraphQl\Exception\GraphQlInputException;
 
 /**
- * @inheritdoc
+ * CatalogGraphQl product option date type.
+ *
+ * @author     Magento Core Team <core@magentocommerce.com>
```
Comment (lenaorobei) [inline app/code/Magento/CatalogGraphQl/Model/Product/Option/DateType.php]:
```suggestion 
```

**Comment by lenaorobei** [comment]
We need to cover this case with `api-functional` test. Please see `\Magento\GraphQl\Quote\AddSimpleProductWithCustomOptionsToCartTest` for example, I noticed that we add to cart with `2019-12-02 20:10:05`, but cart item response is in `Dec 2, 2019` format.

**Comment by lenaorobei** [comment]
Closing this PR since no updates > 3 weeks.
@sedonik feel free to reopen if you want to continue working on it.

**Comment by sedonik** [comment]
Hi @lenaorobei,
Covered with api-functional Tests.
Could you check, please?

**Comment by lenaorobei** [comment]
@magento run WebAPI Tests build

**Comment by TomashKhamlai** [comment]
Hi, @sedonik. Thank you for your contribution. I made some manual checks and found that the input data format is not the same as the output data format.
Request:
```graphql
mutation {
  addSimpleProductsToCart(
    input: {
      cart_id: "XNpDkymvFOQBmxyDExJmgJdEZvjtFubt"
      cart_items: {
        data: {
          quantity: 1
          sku: "simple-product-date"
        }
        customizable_options: [
          {
            id: 2
            value_string: "01-01-01 12:00:02"
          }
          {
            id: 3
            value_string: "01-01-01 00:01:03"
          }
          {
            id: 4
            value_string: "01-01-01 13:00:04"
          }
          {
            id: 5
            value_string: "01-01-01 14:00:05"
          }
        ]
      }
    }
  ) {
    cart {
      items {
        id
        quantity
        product {
          sku
          stock_status
        }
        ... on SimpleCartItem {
          customizable_options {
            id
            label
            values {
              id
              label
              value
            }
          }
        }
      }
    }
  }
}
```
Response:
```graphql
{
  "data": {
    "addSimpleProductsToCart": {
      "cart": {
        "items": [
          {
            "id": "3",
            "quantity": 4,
            "product": {
              "sku": "simple-product-date",
              "stock_status": "IN_STOCK"
            },
            "customizable_options": [
              {
                "id": 2,
                "label": "Date",
                "values": [
                  {
                    "id": 11,
                    "label": "",
                    "value": "Jan 1, 2001"
                  }
                ]
              },
              {
                "id": 3,
                "label": "DateTime",
                "values": [
                  {
                    "id": 12,
                    "label": "",
                    "value": "1/1/01, 12:01 AM"
                  }
                ]
              },
              {
                "id": 4,
                "label": "DateOptional",
                "values": [
                  {
                    "id": 13,
                    "label": "",
                    "value": "Jan 1, 2001"
                  }
                ]
              },
              {
                "id": 5,
                "label": "DateTimeOptional",
                "values": [
                  {
                    "id": 14,
                    "label": "",
                    "value": "1/2/01, 2:00 AM"
                  }
                ]
              }
            ]
          }
        ]
      }
    }
  }
}
```
Could you fix it?

There is also one little problem that I noticed today. Why 14:00:00 is formatted to 02:00AM?
I expect it to be 02:00PM.

**Comment by TomashKhamlai** [comment]
After some conversation, we decided to ask you only to fix the tests. Other problems will be discussed in separate tickets. I will link them here to the original issue.
Please merge the latest mainline, fix tests and this PR is going to be accepted.

**Comment by sedonik** [comment]
Hi @lenaorobei, @TomashKhamlai ,
After merging the branch with upstream, the build was failed but there are no failings in reports. 
Please advice.

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-6077](https://jira.corp.magento.com/browse/ENGCOM-6077) has been created to process this Pull Request

**Pull request merged**